java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -686,9 +740,14 @@ public long getOldestRecordTimeInSeconds() {
         
         long id = messageNumber.getAndIncrement();
         SettableFuture<List<Metric>> f = SettableFuture.create();
-        SettableFutureTracker futureTracker = new SettableFutureTracker(f, Instant.now());
-        futures.put(id, futureTracker);
-        oldestFutureTrackerHeap.add(futureTracker);
+        FutureTask<String> task = null;
+        if(config.getFutureTimeoutInMillis() > 0) {
+            task = new FutureTask(new FutureTimeoutRunnableTask(id), "TimedOut");
+            futureTimeoutExecutor.schedule(task, config.getFutureTimeoutInMillis(), TimeUnit.MILLISECONDS);
```

Comment:
I think i am doing pretty much the same thing. From my understanding we cannot replicate the exact same logic since we dont have a trigger for the getAsync method since our only trigger relies on the message received from native layer which in itself is something we are trying to timeout. I think we can discuss this more if you feel we can simplify this more.